File: data/readable_format/fortran/file2293.f90
PROGRAM TEST_ESSAI
  IMPLICIT NONE

  double precision :: APX
  double precision :: AKPX
  integer :: IND
  logical :: ERREUR

  integer :: I
  double precision :: epsilon
  double precision, dimension(28) :: AKP, AP

  ! Initialize variables
  epsilon = 1.0d-12
  IND = 1
  APX = 1.5d0
  ERREUR = .FALSE.

  ! Populate AKP and AP with some test data
  DO I = 1, 28
    AKP(I) = I * 1.0d0
    AP(I) = I * 1.0d0
  END DO

  ! Call the ESSAI subroutine
  CALL ESSAI(APX, AKPX, IND, ERREUR, AKP, AP, epsilon)

  ! Print the results
  IF (ERREUR) THEN
    PRINT *, 'Error: APX is less than 0.'
  ELSE
    PRINT *, 'AKPX: ', AKPX
  END IF

CONTAINS

  SUBROUTINE ESSAI(APX, AKPX, IND, ERREUR, AKP, AP, epsilon)
    double precision, intent(in) :: APX
    double precision, intent(out) :: AKPX
    integer, intent(in) :: IND
    logical, intent(out) :: ERREUR
    double precision, dimension(28), intent(in) :: AKP, AP
    double precision, intent(in) :: epsilon

    integer :: I, IM
    logical :: FIN_TRAITEMENT

    ERREUR = .FALSE.

    IF (IND > 0) THEN
      IF (APX < 0.d0) THEN
        ERREUR = .TRUE.
        RETURN
      END IF
      FIN_TRAITEMENT = .FALSE.
      I = 1
      DO WHILE ((I <= 28) .AND. (.NOT. FIN_TRAITEMENT))

        IF (ABS(APX-AP(I)) <= epsilon) THEN
          AKPX = AKP(I)
          FIN_TRAITEMENT = .TRUE.
        ELSE IF (APX <= AP(I)) THEN
          IM = I - 1
          AKPX = AKP(IM) + (AKP(I)-AKP(IM)) * (APX-AP(IM)) / (AP(I)-AP(IM))
          FIN_TRAITEMENT = .TRUE.
        ELSE
          I = I + 1
        END IF

      END DO
    END IF
  END SUBROUTINE ESSAI

END PROGRAM TEST_ESSAI

File: data/readable_format/cpp/file2293.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <limits>

void ESSAI(double APX, double& AKPX, int IND, bool& ERREUR, std::vector<double>& AKP, std::vector<double>& AP) {
    static const double epsilon = std::numeric_limits<double>::epsilon();
    
    ERREUR = false;

    if (IND > 0) {
        if (APX < 0.0) {
            ERREUR = true;
            return;
        }

        bool FIN_TRAITEMENT = false;
        int I = 1;
        int IM;

        while (I <= 28 && !FIN_TRAITEMENT) {
            if (std::abs(APX - AP[I - 1]) <= epsilon) {
                AKPX = AKP[I - 1];
                FIN_TRAITEMENT = true;
            } else if (APX <= AP[I - 1]) {
                IM = I - 1;
                AKPX = AKP[IM - 1] + (AKP[I - 1] - AKP[IM - 1]) * (APX - AP[IM - 1]) / (AP[I - 1] - AP[IM - 1]);
                FIN_TRAITEMENT = true;
            } else {
                I = I + 1;
            }
        }
    }
}

int main() {
    double APX;
    double AKPX;
    int IND;
    bool ERREUR;

    std::vector<double> AKP(28);
    std::vector<double> AP(28);

    // Initialize APX, AKP, AP and IND with appropriate values

    ESSAI(APX, AKPX, IND, ERREUR, AKP, AP);

    if (ERREUR) {
        std::cerr << "Error: APX is less than 0." << std::endl;
    } else {
        std::cout << "AKPX: " << AKPX << std::endl;
    }

    return 0;
}